File: s1/L6/z2/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int n;
int *position;
bool *bije_wiersz;
bool *bije_przek1;
bool *bije_przek2;
long long sol_count = 0;

void print_sol() {
    sol_count++;
    for(int i = 0; i < n; i++) {
        printf("%d ", position[i] + 1);
    }
    putchar('\n');
}

void ustaw(int i) {
    for (int j = 0; j < n; j++) {
        int idx1 = i + j;
        int idx2 = i - j + (n - 1);
        if(!(bije_wiersz[j] || bije_przek1[idx1] || bije_przek2[idx2])) {
            position[i] = j;
            bije_wiersz[j] = bije_przek1[idx1] = bije_przek2[idx2] = true;

            if(i < n - 1) ustaw(i + 1);
            else print_sol();

            position[i] = -1;
            bije_wiersz[j] = bije_przek1[idx1] = bije_przek2[idx2] = false;
        }
    }
}

int main(int argc, char *argv[]) {
    if(argc < 2) return 1;
    if(sscanf(argv[1], "%d", &n) != 1) return 1;

    position = malloc(n * sizeof(int));
    bije_wiersz = malloc(n * sizeof(bool));
    bije_przek1 = malloc((2 * n - 1) * sizeof(bool));
    bije_przek2 = malloc((2 * n - 1) * sizeof(bool));

    if(!position || !bije_wiersz || !bije_przek1 || !bije_przek2) {
        fprintf(stderr, "malloc error\n");
        return 1;
    }

    for (int i = 0; i < n; i++) position[i] = -1;
    for (int i = 0; i < n; i++) bije_wiersz[i] = false;
    for (int i = 0; i < 2 * n - 1; i++) bije_przek1[i] = bije_przek2[i] = false;

    ustaw(0);

    printf("Num of solutions: %lld\n", sol_count);

    free(position);
    free(bije_wiersz);
    free(bije_przek1);
    free(bije_przek2);
    return 0;
}

File: s1/L6/z2/python/main.py
import sys

n = 0
position = []
bije_wiersz = []
bije_przek1 = []
bije_przek2 = []
sol_count = 0

def print_sol():
    global n 
    global position 
    global bije_wiersz
    global bije_przek1
    global bije_przek2
    global sol_count
    sol_count += 1
    for i in range(0, n):
        print(f"{position[i] + 1}", end = " ")
    print("")

def ustaw(i):
    global n 
    global position 
    global bije_wiersz
    global bije_przek1
    global bije_przek2
    global sol_count
    for j in range(0, n):
        idx1 = i + j
        idx2 = i - j + (n - 1)
        if False == (bije_wiersz[j] or bije_przek1[idx1] or bije_przek2[idx2]):
            position[i] = j
            bije_wiersz[j] = bije_przek1[idx1] = bije_przek2[idx2] = True

            if(i < n - 1):
                ustaw(i + 1)
            else:
                print_sol()
            position[i] = -1
            bije_wiersz[j] = bije_przek1[idx1] = bije_przek2[idx2] = False


def init_list(l, size, init):
    for i in range(0, size):
        l.append(init)

def main():
    if len(sys.argv) < 2:
        return 1
    global n 
    global position 
    global bije_wiersz
    global bije_przek1
    global bije_przek2
    global sol_count
    n = int(sys.argv[1])
    init_list(position, n, -1)
    init_list(bije_wiersz, n, False)
    init_list(bije_przek1, 2*n - 1, False)
    init_list(bije_przek2, 2*n - 1, False)
    ustaw(0)
    print(f"Num of sols: {sol_count}")

if __name__ == "__main__":
    print(f"return {main()}")
